Refactor several complex methods and add `extra_attrs` to `.extract_words(...)`

jsvine opened the pull request:
This commit largely focuses on refactoring a few previously-complex methods. In the process, however, it also adds support for the long-requested ability to group word-characters on (and convey information about) extra attributes (such as `fontname` and `size`) (#28).

See CHANGELOG.md for more detailed list of changes.

jsvine committed:
Reduce McCabe complexity of convert.serialize(...)

jsvine committed:
Remove reference/access to .figures

Per diagram and explanation in [1], LTFigure is just a layout-estimated
collation of LTCurve objects. For consistency with the rest of
pdfplumber, removing it from the list of parsed/accessible objects,
although we still do process the curves within each figure.

[1] https://github.com/pdfminer/pdfminer.six/blob/develop/docs/source/topic/converting_pdf_to_text.rst

jsvine committed:
Refactor/simplify Page.process_objects

This commit reworks the Page.process_objects method to be less complex
and more readable. Many of the changes stem from a realization that
pdfminer.six already stores most object attributes in the desired type;
we only need to convert those requiring decimalization.

This results in one notable change to the output, which is that
`upright` is no longer converted from a bool to an int. The necessary
downstream changes are reflected in utils.extract_words and convert.py.

Overall the changes result in no substantial change to performance; it
is possible that the code runs ever-so-slightly faster, but to a barely
detectable degree.

jsvine committed:
Remove test_issues.py:test_pr_77 to speed up tests

The test, responding to PR #77, was particularly slow and the
accompanying PDF somewhat large. Now that the fixes to the issues
identified in the PR are fairly tightly integrated into pdfplumber, we
can remove the tests with reasonable confidence, speeding up development
time.

jsvine committed:
Speed up test_ca_warn_report:test_objects

The same goals can be acheived by just testing the first page, instead
of the entire PDF, trimming approximately 20% off the total test suite's
execution time, although YMMV.

jsvine committed:
Refactor .extract_words and allow attrib-grouping

This commit refactors and hopefully makes clearer the logic in
utils.extract_words. It also adds a new parameter, `extra_attrs`, which
allows the user to pass a list of attributes on which to group all
characters.

For instance, passing `extra_attrs=["fontname", "size"]` will not allow
characters with different font names or sizes to become part of the same
word. As a benefit, those resulting word dicts will contain `"fontname"`
and `"size"` attributes — providing a long-requested feature (cf. issue

jsvine committed:
Update CHANGELOG.md

jsvine committed:
Reformat page.py (per newer version of psf/black)

samkit-jain commented on CHANGELOG.md:
```diff
@@ -2,6 +2,19 @@
 
+## [0.5.24] — Unreleased
+### Added
+- Added `extra_attrs=[...]` parameter to `.extract_text(...)` (c8b200) (#28)
```
Can use hyperlinks for commit hash and issue number as it would make it easier to jump to those.

jsvine replied:
Good call. I had hoped GitHub would do that automatically, but it appears not. Now added in 5d51940.

samkit-jain replied:
Missed adding the `/commit` part in the URL which resulted in a 404 when opening the links. Fixed in https://github.com/jsvine/pdfplumber/pull/260/commits/cb924346bb5c144251dfa683bc2863f6e2237a34

samkit-jain commented on pdfplumber/utils.py:
```diff
@@ -205,86 +205,103 @@ def bbox_to_rect(bbox):
     words = []
-    for char in to_list(chars):
-        chars_by_upright[char.get("upright", 1)].append(char)
+    grouped = itertools.groupby(chars, itemgetter("upright", *extra_attrs))
```
`itertools.groupby()` expects the iterable to be sorted. Should that matter here? I think so not.

jsvine replied:
This small-but-smart question led me more broadly to question assumptions about sort-order in `.extract_words`, and led me to (a) rewriting the logic again, in a hopefully more robust and clearly-defined way, and (b) adding a `use_text_flow` parameter. (Open to ideas for other names for that.) More details below, in 84f9181.

jsvine replied:
And, just to follow up on the original question: In order to make the order of the words returned by `.extract_words` predictable and consistent (top-down, and then either left-right or right-left, depending on `horizontal_ltr`), I don't think we want to sort by `extra_attrs`... but I'm open to other perspectives on this.

samkit-jain commented on README.md:
```diff
@@ -100,7 +100,7 @@ The `pdfplumber.Page` class is at the core of `pdfplumber`. Most things you'll d
 |`.filter(test_function)`| Returns a version of the page with only the `.objects` for which `test_function(obj)` returns `True`.|
 |`.extract_text(x_tolerance=3, y_tolerance=3)`| Collates all of the page's character objects into a single string. Adds spaces where the difference between the `x1` of one character and the `x0` of the next is greater than `x_tolerance`. Adds newline characters where the difference between the `doctop` of one character and the `doctop` of the next is greater than `y_tolerance`.|
-|`.extract_words(x_tolerance=3, y_tolerance=3, horizontal_ltr=True, vertical_ttb=True)`| Returns a list of all word-looking things and their bounding boxes. Words are considered to be sequences of characters where (for "upright" characters) the difference between the `x1` of one character and the `x0` of the next is less than or equal to `x_tolerance` *and* where the `doctop` of one character and ...
+|`.extract_words(x_tolerance=3, y_tolerance=3, keep_blank_chars=False, horizontal_ltr=True, vertical_ttb=True, extra_attrs=[])`| Returns a list of all word-looking things and their bounding boxes. Words are considered to be sequences of characters where (for "upright" characters) the difference between the `x1` of one character and the `x0` of the next is less than or equal to `x_tolerance` *and* ...
```
Can also mention that the valid attributes are the ones mentioned in https://github.com/jsvine/pdfplumber/blob/develop/README.md#char-properties

jsvine replied:
Thanks, you're right that this should be more explicit about what an attribute can be. For concision's sake, though, I've decided just to hyperlink to that subsection, rather than adding more words.

jsvine committed:
[changelog] Hyperlink & standardize commit hashes

h/t @samkit-jain for the suggestion

jsvine committed:
Link `extra_attrs` documentation to list of props

h/t @samkit-jain for the suggestion

jsvine committed:
Rework .extract_words and add use_text_flow param

Taking @samkit-jain's smart question about sort-order into account, this
commit reworks `.extract_words`, making the sorting more explicit, and
adding an option to *not* sort, via `use_text_flow=True`, which follows
the original PDF's text flow, rather than presorting all characters top-down,
left-right.

To do so required rethinking the word-delineation logic,
which in turn motivated a shift toward a more object-oriented approach
(especially to keep track of settings without passing them from function
to function). Now `.extract_text`'s logic is encapsulated in a new class,
`TextExtractor`.

jsvine committed:
Add "direction" attribute to extracted words

"direction" == 1 when upright words go LTR and non-upright words go TTB,
and -1 when the opposite is true. This attribute should help in reusing
the results of `.extract_words` in other functions, so that the user
knows how the extracted word corresponds to the word's layout on the
page.

samkit-jain committed:
[changelog] Fix commit hash hyperlinks
Added hyperlinks for issues and pull request references as well

jsvine reviewed (commented):
Ooof, I really should have caught those hyperlink problems. Thank you!
